PEP 797: Arbitrary Object Immortalization (#4497)
python/peps: peps/pep-0797.rst

PEP: 797
Title: Arbitrary Object Immortalization
Author: Peter Bierma <zintensitydev@gmail.com>
Discussions-To: Pending
Status: Draft
Type: Standards Track
Requires: 683
Created: 16-Jul-2025
Python-Version: 3.15
Post-History: `01-Jul-2025 <https://discuss.python.org/t/97306>`__

> Abstract

Author: Ok, will shorten it. Which parts do you think should be excluded/moved?

> Abstract

Reviewer: Most of it, really. The abstract should generally only be a few paragraphs. It explains primarily what the PEP does, not the motivation or implementation details.

> Abstract

Author: I've now significantly reduced the size of the abstract. Does that look better?

PEP: 797

> Title: Arbitrary Object Immortalization

Author: Hmm, I think the current title is good. How tied are you to this suggestion?

> Title: Arbitrary Object Immortalization

Reviewer 2: The current title reads as if object immortalisation is arbitrary, rather than that it is immortalisation of arbitrary objects: it is unclear if the correct grouping should be "arbitrary (object immortalisation)" or "(arbitrary object) immortalisation". The suggestion was an attempt to clarify, though I'd equally be fine with any other phrasing. It's not my PEP, though, so you'd be entirely within your rights to reject this idea and keep the current title.

A

> Title: Arbitrary Object Immortalization

Author: I've learned from PEP 788 that I'm not good at naming PEPs, so I'm inclined to defer to you.

the magic immortal number. But again, this isn't totally safe; for example, due
to the current string interning implementation, immortalizing a string can cause
crashes. To make matters worse, there's no mechanism in place to deallocate an
object once it has become immortal; this is limiting for potential optimizations
using immortality, because immortal objects stored on the heap (and the memory
that they own) are leaked.

What Problem Does Immortality Solve?
************************************

Currently, the main way to share objects between subinterpreters is via
serialization. This works OK, but can be limited for many objects.
Immortality is a necessary precursor to directly sharing objects across multiple
interpreters, as it provides a way to make reference counting thread-safe
without atomic operations on the reference count field (which has proven to
drastically hurt performance).

Subinterpreters aside, immortality is just generally a powerful tool,
especially for CPython maintainers. The authors of this PEP do not expect that
immortality will become a common tool for Python code, but the complexity of

> the approach on its own warrants a PEP.

Reviewer 2: Do you have a rough sense of how much larger? During review it did occur to me that much of this PEP could just be handled on the issue tracker, in a sense it currently functions more as 'Informational' than 'Standards Track' (unless I've misunderstood the extent of the user-observable behavioural changes). It might make sense to bring the shared object proxies into this PEP, we can then remove some of the overly detailed stuff.

A

> the approach on its own warrants a PEP.

Author: Well, it'll definitely be quite large, as it involves combining two very complex proposals into one. How about this: we just do one PEP about shared object proxies (and keep all the immortality stuff as an implementation detail), and then publish an informational PEP about the new APIs and implications. Does that sound good to you?

by the interpreter's GIL, meaning that trying to modify the reference count of
an object that belongs to another interpreter could cause data races.

This isn't a problem on the free-threaded build as reference count operations
are atomic, but subinterpreters are supported on the GIL-enabled build. An
alternative option could be to make reference counting also atomic on the
GIL-enabled build, but that has been shown to damage single-threaded performance
or compromise compatibility with the stable ABI.

Immortality is a clever way of fixing this problem, because by avoiding
reference counting, there is no risk of data races. In fact, both
subinterpreters and free-threading already use immortality for thread-safety
on some common objects, such as ``None``, ``True``, or ``False``.

Lastly, immortality is also faster than deferred reference counting
(:c:func:`PyUnstable_Object_EnableDeferredRefcount`), which is used for
mitigating reference count contention on some objects on the free-threaded build.
It would be nice to allow users (likely large C API wrappers) to take advantage
of that speedup.

Deallocators Rely On the Interpreter State
------------------------------------------

> Another, less intuitive, danger of reference counting is that when the
> object's reference count reaches zero, it immediately tries to get freed.

Reviewer 2: ```suggestion
Another, less intuitive, danger of reference counting is that when the
object's reference count reaches zero, it immediately tries to deallocate itself.
```

In doing so, an object will release its memory back to CPython's object allocator
(typically via :c:func:`PyObject_GC_Del` or :c:func:`PyObject_Free`). This
allocator is per-interpreter, so it is unsafe to try to release an object's
memory in an interpreter different from the one that created it.

Immortality Has No Impact on Reference Counting Performance
-----------------------------------------------------------

If the primary issue with cross-interpreter reference counting is thread-safety,
why not make reference counting atomic? Unfortunately, atomic operations are
slower than their non-atomic counterparts, so this would hurt overall
performance. :pep:`703` managed to solve this issue using biased reference
counting, splitting the reference count field into two, but at the cost of
breaking the stable C API. Atomic reference counting on its own also doesn't

> solve the per-interpreter deallocator issue as mentioned previously.

Reviewer 2: You could phrase this as a rejected idea; to make refcount ops atomic.

Immortality is a Powerful Tool to Wield
***************************************

In general, immortal objects seem to have a lot of applications that aren't
immediately clear to CPython maintainers. For example, `Nanobind used immortal objects`_
in its original :term:`free threading` implementation. So, it seems like immortality
has interesting use-cases in third-party libraries.

> .. _Nanobind used immortal objects: https://github.com/wjakob/nanobind/pull/695

Reviewer 2: This should be combined into the section around Line 100, talking about uses in third-party libraries.

Specification
=============

Changes to the Internal API
***************************

Internally, there are two functions for making an object immortal:
``_Py_SetImmortal`` and ``_Py_SetImmortalUntracked``.
Because the interpreter might want to do its own things with immortal
objects, either now or in the future, the interpreter state should yield
control of a user-defined immortal object to the internal API. So,
the existing internal functions will gain an extra case to remove an object
from the user-defined immortals, so the interpreter won't try and manually
free it later. A new internal function, ``_Py_SetImmortalKnown``, will be
used instead for places where it's certain that objects could not have

> been made immortal by a user.

Reviewer 2: I feel like this is too much prose for the specification, the discussion over the current internal functions etc. should probably go into *Rationale*. Instead, I would have this section say "We introduce a third function for immortalising an object, :c:func`!_Py_SetImmortalKnown`. This will be used in places where ...".

Changes to String Interning
***************************

Currently, the string interning implementation is reliant

> on the idea that it controls when a string will become immortal.

Reviewer 2: I'd avoid "it" here, the implicit object is the implementation of the string interning algorithm, which isn't immediately obvious on first reading.

Since a user can now immortalize a string on their own, string

> interning has to be made "self-aware," meaning that it has to be

Reviewer 2: Gosh! On a more serious note, perhaps rephrase to note that string interning will need to check if a string is already immortal or etc.

able to dynamically figure out if its immortal or not.

The only place where this needs to change is in
:c:func:`PyUnicode_CHECK_INTERNED`. If the current state is
``SSTATE_INTERNED_MORTAL``, but the object is immortal, then it should

> return ``SSTATE_INTERNED_IMMORTAL`` instead.

Reviewer 2: These details feel a little in-the-weeds, but probably does no harm to keep.

Changes to Object Finalization
******************************

> The C API function, :c:func:`PyObject_CallFinalizer`, must be able to
> de-duplicate calls to immortal objects, even if they're non-GC types.

Reviewer 2: Perhaps "ensure object finalization only occurs once" instead of "de-duplicate"?
```suggestion
The :c:func:`PyObject_CallFinalizer` function must be able to
de-duplicate calls to finalize immortal objects, even if they're non-GC types.
```

Immortalization APIs
********************

> This PEP introduces two new semi-public APIs for making an object immortal:
> :c:func:`PyUnstable_Immortalize` and :func:`sys._immortalize`.

Reviewer 2: They're not semi-public, one is private and the other is unstable. We should have clarity over our internal API contracts.
```suggestion
This PEP introduces two new APIs for making an object immortal:
:c:func:`PyUnstable_Immortalize` and :func:`sys._immortalize`.
```

     object's :c:member:`~PyTypeObject.tp_traverse` slot, if the type can
     contain circular references. This is a general requirement for any garbage
     collected type.
   * All finalization for an object must be done in
     :c:member:`~PyTypeObject.tp_finalize` by
     :c:func:`PyObject_CallFinalizerFromDealloc`.

   .. warning::
      Abuse of this function can result in high memory usage for a program.

.. function:: sys._immortalize(obj)

   Make *obj* an :term:`immortal` object. This means that it will not be deallocated for the
   lifetime of the Python interpreter.

   For objects that have costly modifications to their :term:`reference count`,
   making them :term:`immortal` can help improve performance. For example, on
   the :term:`free-threaded <free threading>` build, making an object
   :term:`immortal` can help mitigate bottlenecks that come from concurrent
   modification of an object's :term:`reference count`.

   .. warning::
      This function should be considered a low-level routine that most users should avoid. Making too many
      objects immortal can result in drastically higher memory usage for a program.

Immortalization Contract

> ************************

Reviewer 2: This topic & idea is introduced in *Specification*. Can you move the bits that already exist today to an earlier section, so that the Specification only needs to detail the changes or strengthening to the contract?

See `Backwards Compatibility`_ for more information.

Reference Counts Cannot be Relied On
------------------------------------

This is already true today, but becomes moreso with this proposal.
Any object that relies on the reference count (typically, in assertions
or in tests) will now break if they are made immortal, because the reference
count of an object will unexpectedly not change.

In one of the 3.14 alphas, this rule about reference counting was
`brought up`_ for clarification by a user of the
limited C API. From that post, it's clear that reference count numbers are
an implementation detail:

    But honestly, expecting specific refcounts is not going to work in the
    future anyway, due to changes in GC implementation and free threading.

In addition, the :term:`documentation <reference count>` explicitly mentions
that reference counts are unstable:

    In CPython, reference counts are not considered to be stable or
    well-defined values; the number of references to an object, and how that
    number is affected by Python code, may be different between versions.

.. _brought up: https://discuss.python.org/t/72006

> Destructors Must Use ``tp_finalize``
> ------------------------------------

Reviewer 2: ```suggestion
Destructors Must Use :c:member:`~PyTypeObject.tp_finalize`
----------------------------------------------------------
```

    only a best practice and not a hard requirement.

Reference Cycles Must be Clearable and Traversable
--------------------------------------------------

In order for immortal object deallocation to be safe, an object that can have
reference cycles must be able to clear those reference cycles via
their :c:member:`~PyTypeObject.tp_clear` slot.

Currently, most extension types that hold a :term:`strong reference` to another
Python object will be able to properly clear references through the garbage
collector. In rare cases, it's possible that a buggy extension type could miss
traversing a strong reference. For immortal object deallocation to work
correctly, Python must be able to identify reference cycles on immortal objects.
Prior to this proposal, reference cycles that aren't traversed by the garbage
collector will just leak, but now if they're immortalized, they
will crash. There's not all that much that can be done about this problem; it's
a bug anyway to have these hidden reference cycles, but it just becomes more of
an issue with this PEP.

Immortal Object Deallocation
****************************

The hardest part of immortalizing arbitrary objects is dealing with

> deallocating them. The "intuitive" way of doing it would be to go

Reviewer 2: This repeats the previous section, "*The hardest part about immortality is the deallocation*". I'd suggest that prose & the problem statement should be moved to *Motivation* & *Rationale*.

For immortal object deallocation, the interpreter needs to temporarily suspend
``free()`` operations on objects so reference cycles won't break things
This algorithm will be known as "deferred memory deletion".

When deferred memory deletion is enabled, calling :c:func:`PyMem_Free`,
:c:func:`PyObject_Free`, or :c:func:`PyObject_GC_Del` will not free the
passed pointer, but instead store the pointer in a deferred memory bank
to be deleted later. The pointer will be kept as completely valid memory
until the deferred memory bank is cleared.

Some objects also use internal freelist APIs instead of usual deallocators,
so those have to be disabled during deferred memory deallocation. Python's
internal freelists are cleared and stored in the deferred memory bank to be
restored later.

After deferred memory deletion has been disabled and the pointers stored in
the deferred memory bank have been deleted, the interpreter must not try
to modify the reference count of any object visible to a user, as it could
have been made immortal and is now deleted. As such, the deferred memory bank
should be cleared at the very end of an interpreter's life.

> Preventing Reference Cycle UAFs
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer 2: ```suggestion
Preventing Reference Cycle Use After Frees
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
